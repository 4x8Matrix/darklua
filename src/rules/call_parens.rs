use crate::nodes::{Arguments, Block, Expression, FunctionCall, StringExpression, TableExpression};
use crate::process::{DefaultVisitorMut, NodeProcessorMut, NodeVisitorMut};
use crate::rules::{Rule, RuleConfigurationError, RuleProperties};

use std::mem;

#[derive(Debug, Clone, Default)]
struct Processor {}

impl NodeProcessorMut for Processor {
    fn process_function_call(&mut self, call: &mut FunctionCall) {
        let new_arguments = match call.mutate_arguments() {
            Arguments::Tuple(expressions) if expressions.len() == 1 => {
                let expression = expressions.iter_mut().next().unwrap();

                match expression {
                    Expression::String(string) => {
                        let mut steal_string = StringExpression::empty();
                        mem::swap(string, &mut steal_string);
                        Some(Arguments::String(steal_string))
                    }
                    Expression::Table(table) => {
                        let mut steal_table = TableExpression::default();
                        mem::swap(table, &mut steal_table);
                        Some(Arguments::Table(steal_table))
                    }
                    _ => None,
                }
            }
            _ => None,
        };

        if let Some(new_arguments) = new_arguments {
            *call.mutate_arguments() = new_arguments;
        }
    }
}

pub const REMOVE_FUNCTION_CALL_PARENS: &'static str = "remove_function_call_parens";

/// A rule that removes parentheses when calling functions with a string or a table.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct RemoveFunctionCallParens {}

impl Rule for RemoveFunctionCallParens {
    fn process(&self, block: &mut Block) {
        let mut processor = Processor::default();
        DefaultVisitorMut::visit_block(block, &mut processor);
    }

    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        for (key, _value) in properties {
            return Err(RuleConfigurationError::UnexpectedProperty(key))
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        REMOVE_FUNCTION_CALL_PARENS
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        RuleProperties::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;

    fn new_rule() -> RemoveFunctionCallParens {
        RemoveFunctionCallParens::default()
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_remove_function_call_parens", rule);
    }
}
