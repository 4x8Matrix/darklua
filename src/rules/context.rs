use crate::nodes::Block;

use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
enum Feature {
    ShallowMerge,
}

/// The intent of this struct is to hold data shared across all rules applied to a file.
#[derive(Debug, Clone, Default)]
pub struct Context {
    features: HashMap<Feature, String>,
}

impl Context {
    pub fn pre_process_block(&mut self, block: &Block) {

    }

    pub fn post_process_block(&self, block: &mut Block) {

    }

    /// Returns an identifier for a shallow merge function
    pub fn request_shallow_merge_identifier(&mut self) -> String {
        if let Some(identifier) = self.features.get(&Feature::ShallowMerge) {
            identifier.to_owned()
        } else {
            let id = "_DARKLUA_SHALLOW_MERGE".to_owned();
            self.features.insert(Feature::ShallowMerge, id.clone());
            id
        }
    }
}
