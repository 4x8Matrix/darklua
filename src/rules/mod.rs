//! A module that contains the different rules that mutates a Lua block.

mod context;
pub use context::Context;

// rules
mod call_parens;
mod compute_expression;
mod convert_lux_to_roact;
mod empty_do;
mod group_local;
mod inject_value;
mod method_def;
mod no_local_function;
mod rename_variables;
mod unused_if_branch;
mod unused_while;

pub use call_parens::*;
pub use compute_expression::*;
pub use convert_lux_to_roact::*;
pub use empty_do::*;
pub use group_local::*;
pub use inject_value::*;
pub use method_def::*;
pub use no_local_function::*;
pub use rename_variables::*;
pub use unused_if_branch::*;
pub use unused_while::*;

use crate::nodes::Block;

use serde::{Serialize, Serializer, Deserialize, Deserializer};
use serde::ser::SerializeMap;
use serde::de::{self, MapAccess, Visitor};
use std::fmt;
use std::str::FromStr;
use std::collections::HashMap;

/// In order to be able to weakly-type the properties of any rule, this enum makes it possible to
/// easily use serde to gather the value associated with a property.
#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum RulePropertyValue {
    Boolean(bool),
    String(String),
    Usize(usize),
    Float(f64),
    StringList(Vec<String>),
    None,
}

/// When implementing the configure method of the Rule trait, the method returns a result that uses
/// this error type.
#[derive(Debug, Clone)]
pub enum RuleConfigurationError {
    /// When a rule gets an unknown property. The string should be the unknown field name.
    UnexpectedProperty(String),
    /// When a rule has a required property. The string should be the field name.
    MissingProperty(String),
    /// When a property is associated with something else than an expected string. The string is
    /// the property name.
    StringExpected(String),
    /// When a property is associated with something else than an expected unsigned number. The
    /// string is the property name.
    UsizeExpected(String),
    /// When a property is associated with something else than an expected list of strings. The
    /// string is the property name.
    StringListExpected(String),
    /// When the value type is invalid. The string is the property name that was given the wrong
    /// value type.
    UnexpectedValueType(String),
}

impl fmt::Display for RuleConfigurationError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use RuleConfigurationError::*;

        match self {
            UnexpectedProperty(property) => write!(f, "unexpected field '{}'", property),
            MissingProperty(property) => write!(f, "missing required field '{}'", property),
            StringExpected(property) => write!(f, "string value expected for field '{}'", property),
            UsizeExpected(property) => write!(f, "unsigned integer expected for field '{}'", property),
            StringListExpected(property) => write!(f, "list of string expected for field '{}'", property),
            UnexpectedValueType(property) => write!(f, "unexpected type for field '{}'", property),
        }
    }
}

pub type RuleProperties = HashMap<String, RulePropertyValue>;

pub type RuleProcessResult = Result<(), Vec<String>>;

/// Defines an interface that will be used to mutate blocks and how to serialize and deserialize
/// the rule configuration.
pub trait Rule: RuleConfiguration {
    /// This method should mutate the given block to apply the rule.
    fn process(&self, block: &mut Block, context: &mut Context) -> RuleProcessResult;
}

pub trait RuleConfiguration {
    /// The rule deserializer will construct the default rule and then send the properties through
    /// this method to modify the behavior of the rule.
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError>;
    /// This method should the unique name of the rule.
    fn get_name(&self) -> &'static str;
    /// For implementing the serialize trait on the Rule trait, this method should return all
    /// properties that differs from their default value.
    fn serialize_to_properties(&self) -> RuleProperties;
}

pub trait FlawlessRule {
    fn flawless_process(&self, block: &mut Block, context: &mut Context);
}

impl<T: FlawlessRule + RuleConfiguration> Rule for T {
    fn process(&self, block: &mut Block, context: &mut Context) -> RuleProcessResult {
        self.flawless_process(block, context);
        Ok(())
    }
}

/// A function to get the default rule stack for darklua. All the rules here must preserve all the
/// functionalities of the original code after being applied. They must guarantee that the
/// processed block will work as much as the original one.
pub fn get_default_rules() -> Vec<Box<dyn Rule>> {
    vec![
        Box::new(ComputeExpression::default()),
        Box::new(RemoveUnusedIfBranch::default()),
        Box::new(RemoveUnusedWhile::default()),
        Box::new(RemoveEmptyDo::default()),
        Box::new(RemoveMethodDefinition::default()),
        Box::new(ConvertLocalFunctionToAssign::default()),
        Box::new(GroupLocalAssignment::default()),
        Box::new(RenameVariables::default()),
        Box::new(RemoveFunctionCallParens::default()),
    ]
}

impl FromStr for Box<dyn Rule> {
    type Err = String;

    fn from_str(string: &str) -> Result<Self, Self::Err> {
        let rule: Box<dyn Rule> = match string {
            COMPUTE_EXPRESSIONS_RULE_NAME => Box::new(ComputeExpression::default()),
            CONVERT_LUX_TO_ROACT_CODE_RULE_NAME => Box::new(ConvertLUXToRoactCode::default()),
            CONVERT_LOCAL_FUNCTION_TO_ASSIGN_RULE_NAME => Box::new(ConvertLocalFunctionToAssign::default()),
            GROUP_LOCAL_ASSIGNMENT => Box::new(GroupLocalAssignment::default()),
            INJECT_GLOBAL_VALUE_RULE_NAME => Box::new(InjectGlobalValue::default()),
            REMOVE_EMPTY_DO_RULE_NAME => Box::new(RemoveEmptyDo::default()),
            REMOVE_FUNCTION_CALL_PARENS => Box::new(RemoveFunctionCallParens::default()),
            REMOVE_METHOD_DEFINITION_RULE_NAME => Box::new(RemoveMethodDefinition::default()),
            REMOVE_UNUSED_IF_BRANCH_RULE_NAME => Box::new(RemoveUnusedIfBranch::default()),
            REMOVE_UNUSED_WHILE_RULE_NAME => Box::new(RemoveUnusedWhile::default()),
            RENAME_VARIABLES_RULE_NAME => Box::new(RenameVariables::default()),
            _ => return Err(format!("invalid rule name: {}", string)),
        };

        Ok(rule)
    }
}

impl Serialize for Box<dyn Rule> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let properties = self.serialize_to_properties();
        let property_count = properties.len();
        let rule_name = self.get_name();

        if property_count == 0 {
            serializer.serialize_str(rule_name)

        } else {
            let mut map = serializer.serialize_map(Some(property_count + 1))?;

            map.serialize_entry("rule", rule_name)?;

            let mut ordered: Vec<(String, RulePropertyValue)> = properties.into_iter().collect();

            ordered.sort_by(|a, b| a.0.cmp(&b.0));

            for (key, value) in ordered {
                map.serialize_entry(&key, &value)?;
            }

            map.end()
        }
    }
}

impl<'de> Deserialize<'de> for Box<dyn Rule> {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Box<dyn Rule>, D::Error> {

        struct StringOrStruct;

        impl<'de> Visitor<'de> for StringOrStruct {
            type Value = Box<dyn Rule>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("rule name or rule object")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E> where E: de::Error {
                let mut rule: Self::Value = FromStr::from_str(value)
                    .map_err(de::Error::custom)?;

                rule.configure(RuleProperties::new())
                    .map_err(de::Error::custom)?;

                Ok(rule)
            }

            fn visit_map<M>(self, mut map: M) -> Result<Self::Value, M::Error> where M: MapAccess<'de> {
                let mut rule_name = None;
                let mut properties = HashMap::new();

                while let Some(key) = map.next_key::<String>()? {
                    match key.as_str() {
                        "rule" => if rule_name.is_none() {
                            rule_name.replace(map.next_value::<String>()?);
                        } else {
                            return Err(de::Error::duplicate_field("rule"))
                        }
                        property => {
                            let value = map.next_value::<RulePropertyValue>()?;

                            if properties.insert(property.to_owned(), value).is_some() {
                                return Err(de::Error::custom(format!("duplicate field {} in rule object", property)))
                            }
                        }
                    }
                }

                if let Some(rule_name) = rule_name {
                    let mut rule: Self::Value = FromStr::from_str(&rule_name)
                        .map_err(de::Error::custom)?;

                    rule.configure(properties).map_err(de::Error::custom)?;

                    Ok(rule)
                } else {
                    Err(de::Error::missing_field("rule"))
                }
            }
        }

        deserializer.deserialize_any(StringOrStruct)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;

    #[test]
    fn snapshot_default_rules() {
        let rules = get_default_rules();

        assert_json_snapshot!("default_rules", rules);
    }
}
