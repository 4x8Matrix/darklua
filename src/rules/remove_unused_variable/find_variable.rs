use crate::nodes::{
    Expression,
    AssignStatement,
    Statement,
    LastStatement,
    LocalFunctionStatement,
    Prefix,
    Variable,
};
use crate::process::{
    NodeProcessor,
    Scope,
};

use std::collections::HashSet;

#[derive(Debug, Clone)]
pub struct FindVariable {
    processing_assignment: bool,
    usage_found: bool,
    identifier: String,
    identifiers: Vec<HashSet<String>>,
}

impl FindVariable {
    pub fn new<S: Into<String>>(identifier: S) -> Self {
        Self {
            processing_assignment: false,
            usage_found: false,
            identifier: identifier.into(),
            identifiers: Vec::new(),
        }
    }

    #[inline]
    pub fn has_found_usage(&self) -> bool {
        self.usage_found
    }

    fn is_identifier_used(&self, identifier: &String) -> bool {
        self.identifiers.iter()
            .any(|set| set.contains(identifier))
    }

    fn insert_identifier(&mut self, identifier: &String) {
        if let Some(set) = self.identifiers.last_mut() {
            set.insert(identifier.clone());
        } else {
            let mut set = HashSet::new();
            set.insert(identifier.clone());
            self.identifiers.push(set);
        }
    }

    fn is_reading_in_variable(&self, variable: &Variable) -> bool {
        match variable {
            Variable::Identifier(_) => false,
            Variable::Field(field) => {
                self.is_reading_in_prefix(field.get_prefix(), true)
            }
            Variable::Index(index) => {
                self.is_reading_in_prefix(index.get_prefix(), true)
                || self.is_reading_in_expression(index.get_index(), false)
            }
        }
    }

    fn is_reading_in_expression(&self, expression: &Expression, is_assignment: bool) -> bool {
        match expression {
            Expression::Identifier(identifier) => {
                if is_assignment {
                    false
                } else {
                    identifier == &self.identifier
                }
            }
            Expression::Field(field) => {
                self.is_reading_in_prefix(field.get_prefix(), is_assignment)
            }
            Expression::Index(index) => {
                self.is_reading_in_prefix(index.get_prefix(), is_assignment)
                || self.is_reading_in_expression(index.get_index(), false)
            }
            Expression::Parenthese(expression) => {
                self.is_reading_in_expression(expression, is_assignment)
            }
            _ => false,
        }
    }

    fn is_reading_in_prefix(&self, prefix: &Prefix, is_assignment: bool) -> bool {
        match prefix {
            Prefix::Identifier(identifier) => {
                if is_assignment {
                    false
                } else {
                    identifier == &self.identifier
                }
            }
            Prefix::Field(field) => {
                self.is_reading_in_prefix(field.get_prefix(), is_assignment)
            }
            Prefix::Index(index) => {
                self.is_reading_in_prefix(index.get_prefix(), is_assignment)
                || self.is_reading_in_expression(index.get_index(), false)
            }
            Prefix::Parenthese(expression) => {
                self.is_reading_in_expression(expression, true)
            }
            Prefix::Call(call) => {
                self.is_reading_in_prefix(call.get_prefix(), false)
            }
        }
    }
}

impl NodeProcessor for FindVariable {
    fn process_variable_expression(&mut self, variable: &String) {
        if !self.usage_found && !self.processing_assignment {
            if variable == &self.identifier && !self.is_identifier_used(variable) {
                self.usage_found = true;
            }
        }
    }

    fn process_assign_statement(&mut self, assign: &AssignStatement) {
        if !self.usage_found {
            self.processing_assignment = true;
            self.usage_found = assign.get_variables().iter()
                .any(|variable| self.is_reading_in_variable(variable));
        }
    }

    fn process_statement(&mut self, _: &Statement) {
        self.processing_assignment = false;
    }

    fn process_last_statement(&mut self, _: &LastStatement) {
        self.processing_assignment = false;
    }
}

impl Scope for FindVariable {
    fn push(&mut self) {
        self.identifiers.push(HashSet::new())
    }

    fn pop(&mut self) {
        self.identifiers.pop();
    }

    fn insert(&mut self, identifier: &String) {
        self.insert_identifier(identifier);
    }

    fn insert_local(&mut self, identifier: &String, _value: Option<&Expression>) {
        self.insert_identifier(identifier);
    }

    fn insert_local_function(&mut self, function: &LocalFunctionStatement) {
        self.insert_identifier(function.get_identifier());
    }
}
