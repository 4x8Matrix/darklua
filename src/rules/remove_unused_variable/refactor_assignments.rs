use crate::nodes::{
    Expression,
    Statement,
};
use crate::process::Evaluator;

#[derive(Debug, PartialEq, Eq)]
pub struct RefactorResult<T> {
    pub statements: Vec<Statement>,
    pub variables: Vec<T>,
    pub values: Vec<Expression>,
}

impl<T> RefactorResult<T> {
    fn empty() -> Self {
        Self {
            statements: Vec::new(),
            variables: Vec::new(),
            values: Vec::new(),
        }
    }

    fn with_statements(mut self, statements: Vec<Statement>) -> Self {
        self.statements = statements;
        self
    }

    fn with_variables(mut self, variables: Vec<T>) -> Self {
        self.variables = variables;
        self
    }

    fn with_values(mut self, values: Vec<Expression>) -> Self {
        self.values = values;
        self
    }
}

fn refactor_remove_all<T>(
    // a vector of variables and a boolean that tells if that variable
    // should be kept and if it contains side effects
    assignments: Vec<(T, bool, bool)>,
    values: Vec<&Expression>,
    evaluator: &Evaluator,
) -> RefactorResult<T> {
    let keep_values: Vec<&Expression> = values.into_iter()
        .filter(|value| evaluator.has_side_effects(value))
        .collect();

    let mut calls_at_front = Vec::new();
    let mut remaining_values = Vec::new();
    let mut only_function_calls = true;

    keep_values.into_iter()
        .for_each(|value| {
            if only_function_calls {
                match value {
                    Expression::Call(call) => {
                        calls_at_front.push(
                            Statement::Call(call.as_ref().clone())
                        );
                    }
                    _ => {
                        only_function_calls = false;
                        remaining_values.push(value.clone());
                    }
                }
            } else {
                remaining_values.push(value.clone());
            }
        });

    if remaining_values.is_empty() {
        RefactorResult::empty()
            .with_statements(calls_at_front)
    } else {
        let (variable, _, _) = assignments.into_iter()
            .next()
            .unwrap();
        RefactorResult::empty()
            .with_statements(calls_at_front)
            .with_variables(vec![variable])
            .with_values(remaining_values)
    }
}

pub fn refactor_assignments<T>(
    // a vector of variables and a boolean that tells if that variable
    // should be kept and if it contains side effects
    assignments: Vec<(T, bool, bool)>,
    values: Vec<&Expression>,
    evaluator: &Evaluator,
) -> RefactorResult<T> {
    if assignments.iter().all(|(_, keep, _side_effect)| !keep) {
        refactor_remove_all(assignments, values, evaluator)
    } else {
        let value_count = values.len();
        let variable_count = assignments.len();

        let mut reversed_values = values.into_iter().rev();
        let reversed_values = reversed_values.by_ref();

        let extra_value_count = if value_count >= variable_count {
            value_count - variable_count
        } else {
            0
        };

        let extra_values: Vec<Expression> = reversed_values.take(extra_value_count)
            .filter_map(|value| {
                if evaluator.has_side_effects(value) {
                    Some((*value).clone())
                } else {
                    None
                }
            })
            .collect();

        let mut variables = assignments.into_iter().rev();

        let variables = variables.by_ref();

        let last_variables_count = 1 + if variable_count >= value_count {
            variable_count - value_count
        } else {
            0
        };
        let last_variables_for_last_value = variables.take(last_variables_count);

        let mut reversed_values_for_assignment = Vec::new();
        let mut reversed_variables: Vec<T> = last_variables_for_last_value
            .skip_while(|(_, keep, _)| !*keep)
            .map(|(variable, _, _)| variable)
            .collect();

        if !reversed_variables.is_empty() {
            if let Some(last_value) = reversed_values.next() {
                reversed_values_for_assignment.push(last_value.clone());
            }
        }

        while let Some(value) = reversed_values.next() {
            let (variable, keep_variable, _has_side_effect) = variables.next()
                .expect("found lower than expected amount of variables");

            if keep_variable || evaluator.has_side_effects(value) {
                reversed_variables.push(variable);
                reversed_values_for_assignment.push(value.clone());
            }
        }

        reversed_variables.reverse();
        reversed_values_for_assignment.reverse();
        reversed_values_for_assignment.extend(extra_values.into_iter());

        RefactorResult {
            statements: Vec::new(),
            variables: reversed_variables,
            values: reversed_values_for_assignment,
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::nodes::FunctionCall;

    #[test]
    fn removes_variable_without_a_value() {
        let evaluator = Evaluator::default();
        let refactor = refactor_assignments(
            vec![("foo", false, false)],
            Vec::new(),
            &evaluator,
        );
        assert_eq!(refactor, RefactorResult::empty());
    }

    #[test]
    fn keeps_variable_if_marked_as_used() {
        let evaluator = Evaluator::default();
        let refactor = refactor_assignments(
            vec![("foo", true, false)],
            Vec::new(),
            &evaluator,
        );
        let expected = RefactorResult::empty()
            .with_variables(vec!["foo"]);
        assert_eq!(refactor, expected);
    }

    #[test]
    fn returns_a_call_statement() {
        let evaluator = Evaluator::default();
        let call = FunctionCall::from_name("print");
        let call_expression = call.clone().into();
        let refactor = refactor_assignments(
            vec![("foo", false, false)],
            vec![&call_expression],
            &evaluator,
        );
        let expected = RefactorResult::empty()
            .with_statements(vec![call.into()]);
        assert_eq!(refactor, expected);
    }

    #[test]
    fn keeps_unused_variable_for_tuple_extraction() {
        let evaluator = Evaluator::default();
        let refactor = refactor_assignments(
            vec![
                ("a", false, false),
                ("b", false, false),
                ("c", true, false),
            ],
            vec![&Expression::True, &Expression::VariableArguments],
            &evaluator,
        );
        let expected = RefactorResult::empty()
            .with_variables(vec!["b", "c"])
            .with_values(vec![Expression::VariableArguments]);
        assert_eq!(refactor, expected);
    }

    #[test]
    fn keeps_extra_values_after_used_variable() {
        let evaluator = Evaluator::default();
        let call = FunctionCall::from_name("print");
        let call_expression = call.clone().into();
        let refactor = refactor_assignments(
            vec![("a", true, false)],
            vec![&Expression::True, &call_expression],
            &evaluator,
        );
        let expected = RefactorResult::empty()
            .with_variables(vec!["a"])
            .with_values(vec![Expression::True, call_expression]);
        assert_eq!(refactor, expected);
    }

    #[test]
    fn keeps_uninitialized_variable() {
        let evaluator = Evaluator::default();
        let refactor = refactor_assignments(
            vec![("a", true, false)],
            Vec::new(),
            &evaluator,
        );
        let expected = RefactorResult::empty()
            .with_variables(vec!["a"]);
        assert_eq!(refactor, expected);
    }
}
