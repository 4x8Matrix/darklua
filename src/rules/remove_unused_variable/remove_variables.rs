use crate::nodes::{
    Block,
    AssignStatement,
    DoStatement,
    Expression,
    LocalFunctionStatement,
    Statement,
    Variable,
};
use crate::process::{
    Evaluator,
    NodeProcessorMut,
    Scope,
};
use super::refactor_assignments::refactor_assignments;

use std::collections::HashSet;
use std::iter::FromIterator;

#[derive(Debug, Clone)]
pub struct RemoveVariables {
    variables: HashSet<String>,
    identifiers: Vec<HashSet<String>>,
    evaluator: Evaluator,
}

impl RemoveVariables {
    pub fn new(variables: Vec<String>) -> Self {
        Self {
            variables: HashSet::from_iter(variables.into_iter()),
            identifiers: Vec::new(),
            evaluator: Evaluator::default(),
        }
    }

    fn insert_identifier(&mut self, identifier: &String) {
        if let Some(set) = self.identifiers.last_mut() {
            set.insert(identifier.clone());
        } else {
            let mut set = HashSet::new();
            set.insert(identifier.clone());
            self.identifiers.push(set);
        }
    }

    #[inline]
    fn is_identifier_used(&self, identifier: &String) -> bool {
        self.identifiers.iter()
            .any(|set| set.contains(identifier))
    }

    #[inline]
    fn should_remove_identifier(&self, identifier: &String) -> bool {
        println!("  should remove {:?}", identifier);
        self.variables.contains(identifier) &&
            !self.is_identifier_used(identifier)
    }
}

impl Scope for RemoveVariables {
    fn push(&mut self) {
        self.identifiers.push(HashSet::new())
    }

    fn pop(&mut self) {
        self.identifiers.pop();
    }

    fn insert(&mut self, identifier: &String) {
        self.insert_identifier(identifier);
    }

    fn insert_local(&mut self, identifier: &String, _value: Option<&Expression>) {
        self.insert_identifier(identifier);
    }

    fn insert_local_function(&mut self, function: &LocalFunctionStatement) {
        self.insert_identifier(function.get_identifier());
    }
}

impl NodeProcessorMut for RemoveVariables {
    fn process_statement(&mut self, statement: &mut Statement) {
        let replace_with = match statement {
            Statement::Assign(assign) => {
                let values = assign.get_values();
                let variables = assign.get_variables();
                let assignments: Vec<(Variable, bool)> = variables.iter()
                    .map(|variable| {
                        println!("TEST VARIABLE = {:?}", variable);
                        let should_remove = variable.get_root_identifier()
                            .map(|identifier| self.should_remove_identifier(identifier))
                            .unwrap_or(false);

                        (variable.clone(), !should_remove)
                    })
                    .collect();

                let mut refactor = refactor_assignments(
                    assignments,
                    values.into_iter().collect(),
                    &self.evaluator,
                );

                if refactor.values.is_empty() {
                    if refactor.statements.is_empty() {
                        Some(DoStatement::default().into())
                    } else {
                        if refactor.statements.len() == 1 {
                            Some(refactor.statements.pop().unwrap())
                        } else {
                            Some(
                                DoStatement::new(
                                    Block::new(refactor.statements, None)
                                ).into()
                            )
                        }
                    }
                } else {
                    Some(
                        AssignStatement::new(refactor.variables, refactor.values).into()
                    )
                }
            }
            Statement::Function(function) => {
                let root_name = function.get_name().get_identifier();

                if self.should_remove_identifier(root_name) {
                    Some(DoStatement::default().into())
                } else {
                    None
                }
            }
            _ => None,
        };

        if let Some(replace_statement) = replace_with {
            *statement = replace_statement;
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::nodes::{
        Block,
        AssignStatement,
        FunctionStatement,
        Variable,
    };
    use crate::process::{
        DefaultVisitorMut,
        NodeVisitorMut,
    };

    #[test]
    fn remove_assign_statement() {
        let mut processor = RemoveVariables::new(vec!["foo".to_owned()]);

        let mut block = Block::default()
            .with_statement(AssignStatement::new(
                vec![Variable::new("foo")],
                vec![Expression::Nil],
            ));

        DefaultVisitorMut::visit_block(&mut block, &mut processor);

        let expected_block = Block::default()
            .with_statement(DoStatement::default());
        assert_eq!(block, expected_block);
    }

    #[test]
    fn remove_function_statement() {
        let mut processor = RemoveVariables::new(vec!["foo".to_owned()]);

        let mut block = Block::default()
            .with_statement(FunctionStatement::from_name(
                "foo",
                Block::default(),
            ));

        DefaultVisitorMut::visit_block(&mut block, &mut processor);

        let expected_block = Block::default()
            .with_statement(DoStatement::default());
        assert_eq!(block, expected_block);
    }
}
