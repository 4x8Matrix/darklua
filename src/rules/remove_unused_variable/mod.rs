mod find_variable;
mod refactor_assignments;
mod remove_variables;

use find_variable::FindVariable;
use refactor_assignments::refactor_assignments;
use remove_variables::RemoveVariables;

use crate::nodes::{
    Block,
    DoStatement,
    LastStatement,
    LocalAssignStatement,
    Statement,
};
use crate::process::{
    DefaultVisitorMut,
    Evaluator,
    NodeProcessorMut,
    NodeVisitor,
    NodeVisitorMut,
    ScopeVisitor,
};
use crate::rules::{Rule, RuleConfigurationError, RuleProperties};

#[derive(Debug, Clone, PartialEq, Eq)]
enum FilterResult {
    Keep,
    Remove,
}


#[derive(Debug, Clone, PartialEq, Eq)]
enum Modification {
    Remove(Vec<String>),
    Replace(Statement, Vec<String>),
}

#[derive(Debug, Clone, Default)]
struct UnusedFilter {
    evaluator: Evaluator,
}

impl UnusedFilter {
    fn find_usage<'a, I: Iterator<Item=&'a Statement>>(
        &self,
        find_variable: &mut FindVariable,
        mut next_statements: I,
        last_statement: &Option<LastStatement>
    ) -> FilterResult {
        next_statements
            .find(|statement| {
                ScopeVisitor::visit_statement(statement, find_variable);

                find_variable.has_found_usage()
            })
            .map(|_| FilterResult::Keep)
            .or_else(|| {
                last_statement.as_ref()
                    .filter(|last_statement| {
                        ScopeVisitor::visit_last_statement(last_statement, find_variable);

                        find_variable.has_found_usage()
                    })
                    .map(|_| FilterResult::Keep)
            })
            .unwrap_or(FilterResult::Remove)
    }
}

impl NodeProcessorMut for UnusedFilter {
    fn process_block(&mut self, block: &mut Block) {
        let (statements, last_statement) = block.mutate_all_statements();

        for i in (0..statements.len()).rev() {
            if let Some(modification) = {
                let current = statements.get(i).unwrap();
                let next_statements = statements.iter().skip(i + 1);

                match current {
                    Statement::LocalFunction(function) => {
                        let mut find_function = FindVariable::new(function.get_identifier());

                        match self.find_usage(&mut find_function, next_statements, last_statement) {
                            FilterResult::Remove => {
                                Some(Modification::Remove(vec![
                                    function.get_identifier().to_owned(),
                                ]))
                            }
                            _ => None,
                        }
                    }
                    Statement::LocalAssign(assign) => {
                        let values = assign.get_values();
                        let assignments: Vec<(String, bool)> = assign
                            .get_variables()
                            .iter()
                            .map(|identifier| {
                                let mut find_identifier = FindVariable::new(identifier);

                                let usage = self.find_usage(
                                    &mut find_identifier,
                                    next_statements.clone(),
                                    last_statement,
                                );

                                (identifier.clone(), usage == FilterResult::Keep)
                            })
                            .collect();
                        let removed_variable_names = assignments.iter()
                            .filter_map(|(identifier, keep)| {
                                if !keep {
                                    Some(identifier.clone())
                                } else {
                                    None
                                }
                            })
                            .collect();

                        let mut refactor = refactor_assignments(
                            assignments,
                            values.into_iter().collect(),
                            &self.evaluator,
                        );

                        if refactor.variables.is_empty() {
                            if refactor.statements.is_empty() {
                                Some(Modification::Remove(removed_variable_names))
                            } else {
                                if refactor.statements.len() == 1 {
                                    Some(Modification::Replace(
                                        refactor.statements.pop().unwrap(),
                                        removed_variable_names,
                                    ))
                                } else {
                                    Some(Modification::Replace(
                                        DoStatement::new(
                                            Block::new(refactor.statements, None)
                                        ).into(),
                                        removed_variable_names,
                                    ))
                                }
                            }
                        } else {
                            Some(Modification::Replace(
                                LocalAssignStatement::new(
                                    refactor.variables,
                                    refactor.values,
                                ).into(),
                                removed_variable_names,
                            ))
                        }
                    }
                    _ => None,
                }
            } {
                let (variables, remove_from) = match modification {
                    Modification::Remove(variables) => {
                        statements.remove(i);
                        (variables, i)
                    }
                    Modification::Replace(new_statement, variables) => {
                        *statements.get_mut(i).unwrap() = new_statement;
                        (variables, i + 1)
                    }
                };
                println!("remove variables {:?}", variables);

                if variables.is_empty() {
                    continue
                }

                for i in (remove_from..statements.len()).rev() {
                    if let Some(statement) = statements.get_mut(i) {
                        println!("test statement {:?}", statement);
                        let mut remove_variables = RemoveVariables::new(variables.clone());
                        DefaultVisitorMut::visit_statement(statement, &mut remove_variables);

                        let remove_statement = match statement {
                            Statement::Do(do_statement) => do_statement.is_empty(),
                            _ => false,
                        };

                        if remove_statement {
                            statements.remove(i);
                        }
                    } else {
                        break
                    }
                }
            }
        }
    }
}

pub const REMOVE_UNUSED_VARIABLE_RULE_NAME: &'static str = "remove_unused_variable";

/// A rule that removes unused variables.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct RemoveUnusedVariable {}

impl Rule for RemoveUnusedVariable {
    fn process(&self, block: &mut Block) {
        let mut processor = UnusedFilter::default();
        DefaultVisitorMut::visit_block(block, &mut processor);
    }

    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        for (key, _value) in properties {
            return Err(RuleConfigurationError::UnexpectedProperty(key))
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        REMOVE_UNUSED_VARIABLE_RULE_NAME
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        RuleProperties::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;

    fn new_rule() -> RemoveUnusedVariable {
        RemoveUnusedVariable::default()
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_remove_unused_variable", rule);
    }
}
