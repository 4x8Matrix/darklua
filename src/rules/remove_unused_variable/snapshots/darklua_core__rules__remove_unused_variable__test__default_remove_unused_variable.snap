---
source: src/rules/remove_unused_variable/mod.rs
expression: rule

---
"remove_unused_variable"
